unittests/finality_test_cluster.hpp

```diff
@@ -1,116 +1,283 @@
+      if (config.transition_to_savanna) {
+         // transition to Savanna
+         // ---------------------
+         fin_policy_0 = node0.finkeys.transition_to_savanna([&](const signed_block_ptr& b) {
+            for (size_t i=1; i<nodes.size(); ++i)
+               nodes[i].push_block(b);
+            process_votes(1, num_nodes - 1);
+         });
 
-   // send node2's latest vote
-   eosio::chain::vote_status process_node2_vote(vote_mode mode = vote_mode::strong);
-   
-   // returns true if node0's LIB has advanced
-   bool node0_lib_advancing();
+         // at this point, node0 has a QC to include in next block.
+         // Produce that block and push it, but don't process votes so that
+         // we don't start with an existing QC
+         // ---------------------------------------------------------------
+         produce_and_push_block();
 
-   // returns true if node1's LIB has advanced
-   bool node1_lib_advancing();
+         // reset votes and saved lib, so that each test starts in a clean slate
+         // --------------------------------------------------------------------
+         clear_votes_and_reset_lib();
+      }
+   }
+
+   // node0 produces a block and pushes it to node1 and node2
+   signed_block_ptr produce_and_push_block() {
+      auto b = node0.produce_block();
+      for (size_t i=1; i<nodes.size(); ++i)
+         nodes[i].push_block(b);
+      return b;
+   }
 
-   // returns true if node2's LIB has advanced
-   bool node2_lib_advancing();
+   // Produces and propagate finality votes for `block_count` blocks.
+   signed_block_ptr produce_blocks(uint32_t blocks_count)  {
+      signed_block_ptr b;
+      for (uint32_t i=0; i<blocks_count; ++i) {
+         b = produce_and_push_block();
+         process_votes(1, num_nodes - 1);
+      }
+      return b;
+   }
 
    // Produces a number of blocks and returns true if LIB is advancing.
    // This function can be only used at the end of a test as it clears
    // node1_votes and node2_votes when starting.
-   bool produce_blocks_and_verify_lib_advancing();
+   bool produce_blocks_and_verify_lib_advancing() {
+      // start from fresh
+      clear_votes_and_reset_lib();
 
-   // Produces and propagate finality votes block_count blocks.
-   void produce_blocks(uint32_t blocks_count);
+      produce_and_push_block();
+      for (auto i = 0; i < 3; ++i) {
+         process_votes(1, num_needed_for_quorum);
+         produce_and_push_block();
+         if (lib_advancing() < num_nodes)
+            return false;
+      }
 
-   // Intentionally corrupt node1's vote's block_id and save the original vote
-   void node1_corrupt_vote_block_id();
+      return true;
+   }
 
-   // Intentionally corrupt node1's vote's finalizer_key and save the original vote
-   void node1_corrupt_vote_finalizer_key();
+   // returns true if lib advanced on all nodes since we last checked
+   size_t lib_advancing() {
```

Comment:
This returns the number of nodes where LIB is advancing, not true or false. It's better to change the name or return type.

Reply:
done.